src/Algebra/Module/Structures.agda

```diff
@@ -0,0 +1,262 @@
+      *ᵣ-identityʳ : RightIdentity 1# *ᵣ
+      *ᵣ-assoc : Associative _*_ *ᵣ
+      *ᵣ-zeroˡ : LeftZero 0ᴹ *ᵣ
+      *ᵣ-distribʳ : *ᵣ DistributesOverʳ +ᴹ
+
+  record IsRightSemimodule (+ᴹ : Op₂ M) (*ᵣ : Opᵣ R M) (0ᴹ : M)
+                           : Set (r ⊔ m ⊔ ℓr ⊔ ℓm) where
+    open R R _≈ᴹ_
+    field
+      +ᴹ-isCommutativeMonoid : IsCommutativeMonoid _≈ᴹ_ +ᴹ 0ᴹ
+      isPrerightSemimodule : IsPrerightSemimodule +ᴹ *ᵣ 0ᴹ
+
+    open IsPrerightSemimodule isPrerightSemimodule public
+
+    open IsCommutativeMonoid +ᴹ-isCommutativeMonoid public
+      using () renaming
+      ( assoc                to +ᴹ-assoc
+      ; comm                 to +ᴹ-comm
+      ; identity             to +ᴹ-identity
+      ; identityʳ            to +ᴹ-identityʳ
+      ; identityˡ            to +ᴹ-identityˡ
+      ; isEquivalence        to M-isEquivalence
+      ; isMagma              to +ᴹ-isMagma
+      ; isMonoid             to +ᴹ-isMonoid
+      ; isPartialEquivalence to M-isPartialEquivalence
+      ; isSemigroup          to +ᴹ-isSemigroup
+      ; refl                 to M-refl
+      ; reflexive            to M-reflexive
+      ; setoid               to M-setoid
+      ; sym                  to M-sym
+      ; trans                to M-trans
+      ; ∙-cong               to +ᴹ-cong
+      ; ∙-congʳ              to +ᴹ-congʳ
+      ; ∙-congˡ              to +ᴹ-congˡ
+      )
+
+module _ (R-semiring : Semiring r ℓr) (S-semiring : Semiring s ℓs)
+  where
+
+  open Semiring R-semiring using () renaming (Carrier to R)
+  open Semiring S-semiring using () renaming (Carrier to S)
+
+  record IsBisemimodule (+ᴹ : Op₂ M) (*ₗ : Opₗ R M) (*ᵣ : Opᵣ S M) (0ᴹ : M)
+                        : Set (r ⊔ s ⊔ m ⊔ ℓr ⊔ ℓs ⊔ ℓm) where
+    field
+      +ᴹ-isCommutativeMonoid : IsCommutativeMonoid _≈ᴹ_ +ᴹ 0ᴹ
+      isPreleftSemimodule : IsPreleftSemimodule R-semiring +ᴹ *ₗ 0ᴹ
+      isPrerightSemimodule : IsPrerightSemimodule S-semiring +ᴹ *ᵣ 0ᴹ
+      *ₗ-*ᵣ-assoc : ∀ r m s → *ᵣ (*ₗ r m) s ≈ᴹ *ₗ r (*ᵣ m s)
+
+    isLeftSemimodule : IsLeftSemimodule R-semiring +ᴹ *ₗ 0ᴹ
+    isLeftSemimodule = record
+      { +ᴹ-isCommutativeMonoid = +ᴹ-isCommutativeMonoid
+      ; isPreleftSemimodule = isPreleftSemimodule
+      }
+
+    isRightSemimodule : IsRightSemimodule S-semiring +ᴹ *ᵣ 0ᴹ
+    isRightSemimodule = record
+      { +ᴹ-isCommutativeMonoid = +ᴹ-isCommutativeMonoid
+      ; isPrerightSemimodule = isPrerightSemimodule
+      }
+
+    open IsLeftSemimodule isLeftSemimodule public
+      hiding (+ᴹ-isCommutativeMonoid; isPreleftSemimodule)
+    open IsPrerightSemimodule isPrerightSemimodule public
+
+module _ (commutativeSemiring : CommutativeSemiring r ℓr) where
+  open CommutativeSemiring commutativeSemiring renaming (Carrier to R)
+
+  record IsSemimodule (+ᴹ : Op₂ M) (*ₗ : Opₗ R M) (*ᵣ : Opᵣ R M) (0ᴹ : M)
+                      : Set (r ⊔ m ⊔ ℓr ⊔ ℓm) where
+    field
+      isBisemimodule : IsBisemimodule semiring semiring +ᴹ *ₗ *ᵣ 0ᴹ
```

Comment:
This one is going to need a comment? Why is a semimodule the same thing as  bisemimodule? Oh because the underlying semiring is commutative?